src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -171,4 +171,66 @@ sealed class BitPat(val value: BigInt, val mask: BigInt, width: Int) extends Sou
   }.mkString
 
   override def toString = s"BitPat($rawString)"
+
+  /**
+    * If a BitPat overlaps with another BitPat, i.e., exists x : UInt that (x === this) && (x === rhs)
+    *
+    * (lhs overlap rhs) iff bitwise: both care (lhs.mask & rhs.mask) => both equal (lhs.value == rhs.value)
+    */
+  def overlap(rhs: BitPat): Boolean = ((mask & rhs.mask) & (value ^ rhs.value)) == 0
+  /**
+    * If a BitPat contains another BitPat
+    *
+    * (lhs contain rhs) iff bitwise: lhs care => rhs also care && equal (~(rhs.mask & (value xnor rhs.value)))
+    */
+  def contain(rhs: BitPat): Boolean = (mask & (~rhs.mask | (value ^ rhs.value))) == 0
+
+  /**
+    * Calculate intersection of two BitPats
+    * @return A Set of BitPat as the result may not be represented by a single BitPat
+    */
+  def intersect(rhs: BitPat): Set[BitPat] = {
```

Comment:
I have bigger concerns about `intersect` and `subtract`. These are extremely useful operations, but generally, they should return the same types as their arguments. I don't love the fact that they have to return `Set[BitPat]`. I know this is because their results cannot be represented by a single `BitPat`, but this makes me question the `BitPat` API. Perhaps there should be a more generic `BitSet` `trait` of which `BitPat` is one concrete implementation.